lib/_http_server.js

```diff
@@ -255,9 +256,18 @@ function writeHead(statusCode, reason, obj) {
 // Docs-only deprecated: DEP0063
 ServerResponse.prototype.writeHeader = ServerResponse.prototype.writeHead;
 
+function Server(options, requestListener) {
+  if (typeof options === 'function') {
+    requestListener = options;
+    options = {};
+  } else if (options == null || typeof options === 'object') {
+    options = options || {};
+  }
+
+  this.IncomingMessage = options.IncomingMessage || IncomingMessage;
+  this.ServerResponse = options.ServerResponse || ServerResponse;
 
-function Server(requestListener) {
-  if (!(this instanceof Server)) return new Server(requestListener);
+  if (!(this instanceof Server)) return new Server(options, requestListener);
```

Comment:
This should remain the first line to avoid executing the same code above twice in the case of no `new`.

Reply:
Don't we need to initialize options in the case of no `new`?

Reply:
@hekike yes but it can be done after that, see `https.js` for example.